src/decompiler-ast.inl

```diff
@@ -187,81 +189,160 @@ struct AST {
         return;
       }
     }
-    NewNode(NodeType::Expr, e.type(), &e, arity.nargs);
+    InsertNode(NodeType::Expr, e.type(), &e, arity.nargs);
   }
 
   void Construct(const ExprList& es, Index nresults, bool return_results) {
-    auto start = stack.size();
-    auto stack_depth_start = stack_depth;
+    auto start = exp_stack.size();
+    auto value_stack_depth_start = value_stack_depth;
+    auto value_stack_in_variables = value_stack_depth;
     bool unreachable = false;
     for (auto& e : es) {
       Construct(e);
       auto arity = mc.GetExprArity(e);
-      stack_depth -= arity.nargs;
-      stack_depth += arity.nreturns;
+      value_stack_depth -= arity.nargs;
+      value_stack_in_variables = std::min(value_stack_in_variables,
+                                         value_stack_depth);
       unreachable = unreachable || arity.unreachable;
-      assert(unreachable || stack_depth >= 0);
-      if (arity.nreturns > 1) {
-        // Multivalue: we "push" everything on to the stack.
-        NewNode(NodeType::PushAll, ExprType::Nop, nullptr, 1);
-        // All values become pops.
-        for (Index i = 0; i < arity.nreturns; i++)
-          NewNode(NodeType::Pop, ExprType::Nop, nullptr, 0);
-        // TODO: can also implement a push_all_but_one that returns the top,
-        // then insert N-1 pops below it. Or have a function that returns N
-        // values direcly to N arguments for when structs are passed on,
-        // etc.
+      assert(unreachable || value_stack_depth >= value_stack_depth_start);
+      value_stack_depth += arity.nreturns;
+      // We maintain the invariant that a value_stack_depth of N is represented
+      // by the last N exp_stack items (each of which returning exactly 1
+      // value), all exp_stack items before it return void ("statements").
+      // In particular for the wasm stack:
+      // - The values between 0 and value_stack_depth_start are part of the
+      //   parent block, and not touched here.
+      // - The values from there up to value_stack_in_variables are variables
+      //   to be used, representing previous statements that flushed the
+      //   stack into variables.
+      // - Values on top of that up to value_stack_depth are exps returning
+      //   a single value.
+      // The code below maintains the above invariants. With this in place
+      // code "falls into place" the way you expect it.
+      if (arity.nreturns != 1) {
+        auto num_vars = value_stack_in_variables - value_stack_depth_start;
+        auto num_vals = value_stack_depth - value_stack_in_variables;
+        auto GenFlushVars = [&](int nargs) {
+          auto& ftv = InsertNode(NodeType::FlushToVars, ExprType::Nop, nullptr,
+                                 nargs);
+          ftv.var_start = flushed_vars;
+          ftv.var_count = num_vals;
+        };
+        auto MoveStatementsBelowVars = [&](size_t amount) {
```

Comment:
I appreciate that this is a lambda that describes what it's doing, rather than a std::rotate inline :+1: